library/core/src/swapping.rs

```diff
@@ -0,0 +1,182 @@
+    bytes: usize,
+) {
+    // SAFETY: Two legal non-overlapping regions can't be bigger than this.
+    // (And they couldn't have made allocations any bigger either anyway.)
+    // FIXME: Would be nice to have a type for this instead of the assume.
+    unsafe { hint::assert_unchecked(bytes < isize::MAX as usize) };
+
+    let mut i = 0;
+    macro_rules! swap_next_n {
+        ($n:expr) => {{
+            let x: *mut mem::MaybeUninit<[u8; $n]> = x.add(i).cast();
+            let y: *mut mem::MaybeUninit<[u8; $n]> = y.add(i).cast();
+            swap_nonoverlapping_aligned_chunk::<ALIGN, [u8; $n]>(
+                x.as_mut_unchecked(),
+                y.as_mut_unchecked(),
+            );
+            i += $n;
+        }};
+    }
+
+    while bytes - i >= MAX_ALIGN {
+        const { assert!(MAX_ALIGN >= ALIGN) };
+        // SAFETY: the const-assert above confirms we're only ever called with
+        // an alignment equal to or smaller than max align, so this is necessarily
+        // aligned, and the while loop ensures there's enough read/write memory.
+        unsafe {
+            swap_next_n!(MAX_ALIGN);
+        }
+    }
+
+    macro_rules! handle_tail {
+        ($($n:literal)+) => {$(
+            if const { $n % ALIGN == 0 } {
+                // Checking this way simplifies the block end to just add+test,
+                // rather than needing extra math before the check.
+                if (bytes & $n) != 0 {
+                    // SAFETY: The above swaps were bigger, so could not have
+                    // impacted the `$n`-relevant bit, so checking `bytes & $n`
+                    // was equivalent to `bytes - i >= $n`, and thus we have
+                    // enough space left to swap another `$n` bytes.
+                    unsafe {
+                        swap_next_n!($n);
+                    }
+                }
+            }
+        )+};
+    }
+    const { assert!(MAX_ALIGN <= 64) };
+    handle_tail!(32 16 8 4 2 1);
+
+    debug_assert_eq!(i, bytes);
+}
+
+/// Swaps the `C` behind `x` and `y` as untyped memory
+///
+/// # Safety
+///
+/// Both `x` and `y` must be aligned to `ALIGN`, in addition to their normal alignment.
+/// They must be readable and writeable for `sizeof(C)` bytes, as usual for `&mut`s.
+///
+/// (The actual instantiations are usually `C = [u8; _]`, so we get the alignment
+/// information from the loads by `assume`ing the passed-in alignment.)
+// Don't let MIR inline this, because we really want it to keep its noalias metadata
+#[rustc_no_mir_inline]
+#[inline]
+unsafe fn swap_nonoverlapping_aligned_chunk<const ALIGN: usize, C>(
+    x: &mut mem::MaybeUninit<C>,
+    y: &mut mem::MaybeUninit<C>,
+) {
+    assert!(size_of::<C>() % ALIGN == 0);
+
+    let x = ptr::from_mut(x);
+    let y = ptr::from_mut(y);
+
+    // SAFETY: One of our preconditions.
+    unsafe {
+        hint::assert_unchecked(x.is_aligned_to(ALIGN));
+        hint::assert_unchecked(y.is_aligned_to(ALIGN));
```

Comment:
Do these actually help? Seems worth commenting on that since experience shows that `assert_unchecked` can often hurt codegen.